state/cluster/badger/mutator.go

```diff
@@ -218,3 +205,74 @@ func (m *MutableState) Extend(block *cluster.Block) error {
 	}
 	return nil
 }
+
+// checkDupeTransactionsInUnfinalizedAncestry checks for duplicate transactions in the un-finalized
+// ancestry of the given block, and returns a list of all duplicates if there are any.
+func (m *MutableState) checkDupeTransactionsInUnfinalizedAncestry(block *cluster.Block, includedTransactions map[flow.Identifier]struct{}, finalHeight uint64) ([]flow.Identifier, error) {
+
+	var duplicateTxIDs []flow.Identifier
+	err := fork.TraverseBackward(m.headers, block.Header.ParentID, func(ancestor *flow.Header) error {
+		payload, err := m.payloads.ByBlockID(ancestor.ID())
+		if err != nil {
+			return fmt.Errorf("could not retrieve ancestor payload: %w", err)
+		}
+
+		for _, tx := range payload.Collection.Transactions {
+			txID := tx.ID()
+			_, duplicated := includedTransactions[txID]
+			if duplicated {
+				duplicateTxIDs = append(duplicateTxIDs, txID)
+			}
+		}
+		return nil
+	}, fork.ExcludingHeight(finalHeight))
+
+	return duplicateTxIDs, err
+}
+
+// checkDupeTransactionsInFinalizedAncestry checks for duplicate transactions in the finalized
+// ancestry, and returns a list of all duplicates if there are any.
+func (m *MutableState) checkDupeTransactionsInFinalizedAncestry(includedTransactions map[flow.Identifier]struct{}, minRefHeight, maxRefHeight uint64) ([]flow.Identifier, error) {
+	var duplicatedTxIDs []flow.Identifier
+
+	// Let E be the global transaction expiry constant, measured in blocks
+	//
+	// 1. a cluster block's reference block height is equal to the lowest reference
+	//    block height of all its constituent transactions
+	// 2. a transaction with reference block height T is eligible for inclusion in
+	//    any collection with reference block height C where C<=T<C+E
+	//
+	// Therefore, to guarantee that we find all possible duplicates, we need to check
+	// all collections with reference block height in the range (C-E,C+E). Since we
+	// know the actual maximum reference height included in this collection, we can
+	// further limit the search range to (C-E,maxRefHeight]
+
+	// the finalized cluster blocks which could possibly contain any conflicting transactions
+	var clusterBlockIDs []flow.Identifier
+	start := minRefHeight - flow.DefaultTransactionExpiry + 1
+	if start > minRefHeight {
+		start = 0 // overflow check
+	}
+	end := maxRefHeight
+	err := m.db.View(operation.LookupClusterBlocksByReferenceHeightRange(start, end, &clusterBlockIDs))
```

Comment:
this will be on a hot path, when every extending a new cluster block, we will look back X finalized cluster blocks, which means a certain finalized blocks might get queried over and over for many times.

Maybe we could optimize by maintaining some cache here.

Reply:
I benchmarked the lookup operation with 6,000 cluster blocks (about 5x what we would expect given the block/collection production rates on Mainnet) and it takes about 1ms. The data is stored in the key, which is generally stored in memory already by Badger. I don't think a separate cache will help much here.